File: backend/core/internal/logger/logger.go
package logger

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"time"
)

type LogLevel string

const (
	LevelDebug LogLevel = "DEBUG"
	LevelInfo  LogLevel = "INFO"
	LevelWarn  LogLevel = "WARN"
	LevelError LogLevel = "ERROR"
)

type LogEntry struct {
	Timestamp   string                 `json:"timestamp"`
	Level       LogLevel               `json:"level"`
	Service     string                 `json:"service"`
	Environment string                 `json:"environment"`
	RequestID   string                 `json:"request_id,omitempty"`
	TraceID     string                 `json:"trace_id,omitempty"`
	Caller      string                 `json:"caller,omitempty"`
	Message     string                 `json:"message"`
	Error       string                 `json:"error,omitempty"`
	Fields      map[string]interface{} `json:"fields,omitempty"`
}

type ctxKey string

const RequestIDKey ctxKey = "request_id"
const TraceIDKey ctxKey = "trace_id"

var (
	ServiceName = "bank-core"
	Environment = "development"
)

func Init(service string, env string) {
	if service != "" {
		ServiceName = service
	}
	if env != "" {
		Environment = env
	}
}

func logMessage(ctx context.Context, level LogLevel, msg string, err error, fields map[string]interface{}) {
	reqID := ""
	traceID := ""
	if ctx != nil {
		if val, ok := ctx.Value(RequestIDKey).(string); ok {
			reqID = val
		}
		if val, ok := ctx.Value(TraceIDKey).(string); ok {
			traceID = val
		}
	}

	caller := ""
	if _, file, line, ok := runtime.Caller(2); ok {
		caller = fmt.Sprintf("%s:%d", file, line)
	}

	entry := LogEntry{
		Timestamp:   time.Now().UTC().Format(time.RFC3339Nano),
		Level:       level,
		Service:     ServiceName,
		Environment: Environment,
		RequestID:   reqID,
		TraceID:     traceID,
		Caller:      caller,
		Message:     msg,
		Fields:      fields,
	}

	if err != nil {
		entry.Error = err.Error()
	}

	jsonBytes, _ := json.Marshal(entry)
	fmt.Fprintln(os.Stdout, string(jsonBytes))
}

func Info(ctx context.Context, msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	logMessage(ctx, LevelInfo, msg, nil, f)
}

func Warn(ctx context.Context, msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	logMessage(ctx, LevelWarn, msg, nil, f)
}

func Error(ctx context.Context, msg string, err error, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	logMessage(ctx, LevelError, msg, err, f)
}

func Debug(ctx context.Context, msg string, fields ...map[string]interface{}) {
	var f map[string]interface{}
	if len(fields) > 0 {
		f = fields[0]
	}
	logMessage(ctx, LevelDebug, msg, nil, f)
}

File: backend/ai-service/app/logger.py
import json
import logging
import sys
import time
from datetime import datetime, timezone
from contextvars import ContextVar
from typing import Any, Dict, Optional

request_id_ctx: ContextVar[Optional[str]] = ContextVar("request_id", default=None)
trace_id_ctx: ContextVar[Optional[str]] = ContextVar("trace_id", default=None)

class StructuredJsonFormatter(logging.Formatter):
    def __init__(self, service_name: str = "bank-ai", environment: str = "development"):
        super().__init__()
        self.service_name = service_name
        self.environment = environment

    def format(self, record: logging.LogRecord) -> str:
        log_obj: Dict[str, Any] = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": record.levelname,
            "service": self.service_name,
            "environment": self.environment,
            "caller": f"{record.filename}:{record.lineno}",
            "message": record.getMessage(),
        }

        req_id = request_id_ctx.get()
        if req_id:
            log_obj["request_id"] = req_id

        trace_id = trace_id_ctx.get()
        if trace_id:
            log_obj["trace_id"] = trace_id

        if hasattr(record, "fields") and isinstance(record.fields, dict):
            log_obj["fields"] = record.fields

        if record.exc_info:
            log_obj["error"] = self.formatException(record.exc_info)

        return json.dumps(log_obj)

def setup_logger(service_name: str = "bank-ai", environment: str = "development") -> logging.Logger:
    logger = logging.getLogger("bank_ai")
    logger.setLevel(logging.INFO)
    logger.handlers.clear()

    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(StructuredJsonFormatter(service_name=service_name, environment=environment))
    logger.addHandler(handler)
    logger.propagate = False
    return logger

app_logger = setup_logger()
